conans/client/build/cmake_flags.py

```diff
@@ -204,14 +204,19 @@ def _cmake_cross_build_defines(self):
         else:  # detect if we are cross building and the system name and version
             if cross_building(self._conanfile):  # We are cross building
                 if os_ != os_build:
-                    if os_:  # the_os is the host (regular setting)
+                    if os_:
                         definitions["CMAKE_SYSTEM_NAME"] = {"Macos": "Darwin",
                                                             "iOS": "Darwin",
                                                             "tvOS": "Darwin",
                                                             "watchOS": "Darwin",
                                                             "Neutrino": "QNX"}.get(os_, os_)
                     else:
                         definitions["CMAKE_SYSTEM_NAME"] = "Generic"
+                elif not os_:
```

Comment:
I would expect something like this here:

```python
if cross_building(self._conanfile):  # We are cross building
    if os_:
        # If we are cross-building, it doesn't matter if it is because OS or ARCH
        definitions["CMAKE_SYSTEM_NAME"] = {...}.get(os_, os_)
    else:
        definitions["CMAKE_SYSTEM_NAME"] = "Generic"  # To avoid the CMake warning for missing *.cmake file
```

I'm sure there is a reason why we didn't have this more straightforward if/else conditions... maybe it is encoded in some test, but a comment here would be nice.

---

This implementation won't satisfy the ARM cross-compilation: `os:host=Linux`, `os:build=Linux`, `arch:host=x86_64`, `arch:build=arm`. According to [CMake docs](https://cmake.org/cmake/help/v3.6/manual/cmake-toolchains.7.html#cross-compiling-for-linux), here we should generate something with:

```
set(CMAKE_SYSTEM_NAME Linux)
```

Reply:
cross-building is returning ``True`` for Windows Visual building to x86 for x64, but seems that adding ``CMAKE_SYSTEM_NAME Windows`` for that case would be problematic, as cmake not really consider it cross-building, as far as I understood. Trying to do something like the proposed solution, I got some tests failing.

I am not saying that what we have is correct, certainly it is incomplete and fragile/buggy, but the risk of large breakings is important here.

Reply:
Ok, probably it is worth to add a `CONAN_V2_MODE` block with a more straightforward implementation (if we think it is better):

```python
if cross_compiling:
    if CONAN_V2_MODE:
        if os_ == os_build and _os == 'Windows':
            # Nothing here?
        else:
            definitions["CMAKE_SYSTEM_NAME"] = {..., None, 'Generic'}.get(os_, os_)
    else:
        # Current logic plus patches.
        ...
```

I'll wait for answer [here](https://github.com/conan-io/conan/issues/7826#issuecomment-704861201), I really want to know why setting it to `Generic` instead of the actual value.

Reply:
This functionality is going to be moved anyway to toolchains, so CONAN_V2_MODE would probably not be necessary there.